lib/class_def_ast.js

```diff
@@ -0,0 +1,340 @@
+// -*- mode: js; indent-tabs-mode: nil; js-basic-offset: 4 -*-
+//
+// This file is part of ThingTalk
+//
+// Copyright 2015-2016 The Board of Trustees of the Leland Stanford Junior University
+//
+// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
+//
+// See COPYING for details
+"use strict";
+
+const adt = require('adt');
+const assert = require('assert');
+
+const Type = require('./type');
+const { prettyprintValue } = require('./prettyprint');
+
+// Class and function definitions
+
+function makeIndex(args) {
+    var index = {};
+    var i = 0;
+    for (var a of args)
+        index[a] = i++;
+    return index;
+}
+
+const ArgDirection = {
+    IN_REQ: 'in req',
+    IN_OPT: 'in opt',
+    OUT: 'out'
+};
+module.exports.ArgDirection = ArgDirection;
+
+class ArgumentDef {
+    constructor(direction, name, type, metadata, annotations) {
+        this.direction = direction;
+        this.is_input = direction !== ArgDirection.OUT;
+        this.required = direction === ArgDirection.IN_REQ;
+        this.name = name;
+        this.type = type;
+        this.metadata = metadata || {};
+        this.annotations = annotations || {};
+    }
+
+    clone() {
+        const metadata = {};
+        Object.assign(metadata, this.metadata);
+        const annotations = {};
+        Object.assign(annotations, this.annotations);
+
+        return new ArgumentDef(this.direction, this.name, this.type,
+            metadata, annotations);
+    }
+
+    toString() {
+        let annotations = '';
+        Object.entries(this.metadata).forEach(([name, value]) => {
+            annotations += ` #_[${name}=${prettyprintValue(value)}]`;
+        });
+        Object.entries(this.annotations).forEach(([name, value]) => {
+            annotations += ` #_[${name}=${prettyprintValue(value)}]`;
```

Comment:
This should just be `#[]`
(we might need better codegen tests...)

Reply:
fixed